tests/libs/util/socket_helper.cpp

```diff
@@ -0,0 +1,326 @@
+        &sender_address_size,
+        &overlapped,
+        nullptr);
+    if (error != 0) {
+        int wsaerr = WSAGetLastError();
+        if (wsaerr != WSA_IO_PENDING)
+            FAIL("WSARecvFrom failed with " << wsaerr);
+    }
+}
+
+void
+_datagram_receiver_socket::get_sender_address(PSOCKADDR& from, int& from_length)
+{
+    from = (PSOCKADDR)&sender_address;
+    from_length = sender_address_size;
+}
+
+_stream_receiver_socket::_stream_receiver_socket(int _sock_type, int _protocol, uint16_t _port)
+    : _receiver_socket{_sock_type, _protocol, _port}, acceptex(nullptr), accept_socket(0),
+      message_length(recv_buffer.size() - 2 * (sizeof(sockaddr_storage) + 16))
+{
+    if ((sock_type != SOCK_STREAM) || (protocol != IPPROTO_TCP))
+        FAIL("stream_socket only supports these combinations (SOCK_STREAM, IPPROTO_TCP)");
+
+    GUID guid = WSAID_ACCEPTEX;
+    uint32_t bytes;
+    int error = WSAIoctl(
+        socket,
+        SIO_GET_EXTENSION_FUNCTION_POINTER,
+        &guid,
+        sizeof(guid),
+        &acceptex,
+        sizeof(acceptex),
+        reinterpret_cast<LPDWORD>(&bytes),
+        NULL,
+        NULL);
+
+    if (error != 0)
+        FAIL("Obtaining AcceptEx function pointer failed with " << WSAGetLastError());
+
+    // Post listen.
+    listen(socket, SOMAXCONN);
+
+    // Create accept socket.
+    accept_socket = WSASocket(AF_INET6, sock_type, protocol, nullptr, 0, WSA_FLAG_OVERLAPPED);
+    uint32_t ipv6_opt = 0;
+    error =
+        setsockopt(accept_socket, IPPROTO_IPV6, IPV6_V6ONLY, reinterpret_cast<const char*>(&ipv6_opt), sizeof(ULONG));
+    if (error != 0)
+        FAIL("Could not enable dual family endpoint on accept socket: " << WSAGetLastError());
+}
+
+_stream_receiver_socket::~_stream_receiver_socket() { closesocket(accept_socket); }
+
+void
+_stream_receiver_socket::post_async_receive()
+{
+    WSABUF wsa_recv_buffer{static_cast<ULONG>(recv_buffer.size()), reinterpret_cast<char*>(recv_buffer.data())};
+
+    // Create an event handle and set up the overlapped structure.
+    overlapped.hEvent = WSACreateEvent();
+    if (overlapped.hEvent == NULL)
+        FAIL("WSACreateEvent failed with error: " << WSAGetLastError());
+
+    // Post an asynchronous receive on the socket.
+    if (!acceptex(
+            socket,        // Listen socket.
+            accept_socket, // Accept socket.
+            recv_buffer.data(),
+            static_cast<DWORD>(message_length),
+            static_cast<DWORD>(sizeof(sockaddr_storage)) + 16,
+            static_cast<DWORD>(sizeof(sockaddr_storage)) + 16,
+            reinterpret_cast<LPDWORD>(&bytes_received),
+            &overlapped)) {
+        int wsaerr = WSAGetLastError();
+        if (wsaerr != WSA_IO_PENDING)
+            FAIL("AcceptEx failed with " << wsaerr);
+    }
+}
+
+void
+_stream_receiver_socket::get_sender_address(PSOCKADDR& from, int& from_length)
+{
+    from = (PSOCKADDR)(recv_buffer.data() + message_length);
```

Comment:
this looks like a bug to me since a SOCKADDR_IN6 is larger than SOCKADDR so this will not copy from_length bytes.

Reply:
There is no copy here. The function returns a pointer to a buffer that stores the remote address.